### ISSUE TITLE
watch: interrupt ongoing task

### ISSUE THREAD

**Issue by stelcheck** [go-task/task#160]
Use case: I have a task that starts a server in development mode. I would like to use `watch` to interrupt the ongoing build task before starting the new one. 

I assume this would be the desired behavior for long-running build tasks as well.

**Comment by andreynering** [comment]
Hi @stelcheck,

So, Task actually kill the previous long running command before running it again, BUT it doesn't wait the previous one to finish before it runs the new one. I think this may be causing the issue you're having.

This is unfortunate, but it's due limitation of the Go `context` package (which is asynchronously). I'll have to study how to improve that.

**Comment by stelcheck** [comment]
That would indeed cause problems, at the very least intermittently.

The issue I am seeing though seems to be a bit different. Either:

1. The subprocess is killed with SIGKILL or another more exotic signal
2. Logs stop being outputted while the child process is terminating

I'd be happy to see if I can fix this myself if you can give me a pointer as to where to look in the code.

**Comment by stelcheck** [comment]
Actually, I am now convinced the task is not killed: once I send a signal to the task command, **all** instances of the application receive it at once.

```plaintext
 task run -w
task: Started watching for tasks: run
INFO[0000] Logger system set                             forceColors=true format=text
DEBU[0000]                         ___
DEBU[0000]     o__        o__     |   |\
DEBU[0000]    /|          /\      |   |X\
DEBU[0000]    / > o        <\     |   |XX\
INFO[0000] Setting up metrics system                     subpath=/metrics
INFO[0000] Setting up discovery system                   address="127.0.0.1:8500" scheme=http
INFO[0000] Setting up HTTP Server system                 address="127.0.0.1:8080" prefix=/
INFO[0000] Goal server is up and running
INFO[0000] Logger system set                             forceColors=true format=text
DEBU[0000]                         ___
DEBU[0000]     o__        o__     |   |\
DEBU[0000]    /|          /\      |   |X\
DEBU[0000]    / > o        <\     |   |XX\
INFO[0000] Setting up metrics system                     subpath=/metrics
INFO[0000] Setting up discovery system                   address="127.0.0.1:8500" scheme=http
INFO[0000] Setting up HTTP Server system                 address="127.0.0.1:8080" prefix=/
INFO[0000] Goal server is up and running
INFO[0000] Logger system set                             forceColors=true format=text
DEBU[0000]                         ___
DEBU[0000]     o__        o__     |   |\
DEBU[0000]    /|          /\      |   |X\
DEBU[0000]    / > o        <\     |   |XX\
INFO[0000] Setting up metrics system                     subpath=/metrics
INFO[0000] Setting up discovery system                   address="127.0.0.1:8500" scheme=http
INFO[0000] Setting up HTTP Server system                 address="127.0.0.1:8080" prefix=/
INFO[0000] Goal server is up and running


INFO[0165] Received signal interrupt, shutting down
INFO[0165] Stopping Goal server
INFO[0153] Received signal interrupt, shutting down
INFO[0153] Stopping Goal server
INFO[0165] Tearing down discovery system
INFO[0165] Tearing down metrics system
INFO[0165] Tearing down logger system
INFO[0153] Tearing down discovery system
INFO[0153] Tearing down metrics system
INFO[0165] Goal server stopped
INFO[0174] Received signal interrupt, shutting down
INFO[0174] Stopping Goal server
INFO[0153] Tearing down logger system
INFO[0153] Goal server stopped
INFO[0174] Tearing down discovery system
INFO[0174] Tearing down metrics system
INFO[0174] Tearing down logger system
INFO[0174] Goal server stopped
task: Failed to run task "run": context canceled
task: Failed to run task "run": context canceled
```

You can see the task has be started 3 times, and when I CTRL-C the three instances get killed altogether.

**Comment by andreynering** [comment]
@stelcheck It's by design that sending a SIGINT to a process also send it to all subprocesses. This is a Unix thing and we don't have control about it.

But restarting a long-running task once a file changes (`--watch`), well, we could definitely do better here, even it's tricky. We should probably get rid of using Go's context package in this specific case (because cancellation is asynchronous) and manually kill all processes.

That'd require some refactor, and maybe even changes to [mvdan/sh](https://github.com/mvdan/sh).

**Comment by andreynering** [comment]
> It's a bit odd to set the context in the Executor struct. As per the context package recommendations, context should always be passed as the first parameter to a function (and then passed along for each sub-call, so it can e.g. be modified/replaced etc).

@smyrman Yeah, I know. I already starting moving it to parameters, but didn't do in all places yet.

**Comment by frederikhors** [comment]
@andreynering maybe an inspiration from https://github.com/cortesi/modd can help?

**Comment by andreynering** [comment]
@frederikhors Thanks for the suggestion.

Idea: Perhaps we could watch for the PID of the previous command to know when it's killed.

**Comment by adamkissvcc** [comment]
Here is another inspiration: https://github.com/githubnemo/CompileDaemon

Watches files, runs a build command, and launches a run command after the build is successful. Has very graceful shutdown handling with -graceful-kill set to true (with force kill after a timeout).

**Comment by sisp** [comment]
I'd also like to run a dev server while watching for changes to rebuild some other dependencies that the dev server doesn't watch itself. +1 for waiting for a graceful shutdown before re-running the ongoing task.

**Comment by TheOddler** [comment]
I just bumped into this, it doesn't seem to kill my (haskell) server. It restarts the command, but then I get a `Network.Socket.bind: resource busy (Address already in use)` and when I use `ps -eaf | grep backend` then it shows the original one still running.

**Comment by emosenkis** [comment]
I'm running into this also: I have tasks that run various build steps followed by several instances of `task --watch` to restart servers or regenerate source code when relevant files change. I keep ending up with `task --watch` and the `gunicorn` process tree that it spawned continuing silently in the background, eating up CPU/battery and preventing the new server from listening on the port. Another `task --watch` that runs a short-lived command also continues when the parent `task` is killed with Ctrl-C even though without any child process.

**Comment by MedAzizTousli** [comment]
+1 on this task.
